[Feature] Logger 모듈 구현

hoBahk opened the pull request:
## 🔗 연결된 이슈 번호
- [#15]


##  📝 작업 내용

**Logger 모듈 작업**
- OSLog 사용
- 로그 레벨 지원(Debug, Info, Warning, Error, Fatal)
- 출처 추적 가능(로그가 발생한 파일, 라인 넘버, 함수)
- Debug 빌드에서만 출력되도록 지원 (#if DEBUG)
- 로그 레벨 별 출력 제어
- 비동기로 구현(메인쓰레드에서 블락 되어 로그 출력이 안되는 일 없도록)
- 필터링/태깅

  

## 📣 그 외 알릴 내용

### 로그 레벨

- Debug(Debug): 개발 중 코드 디버깅 시 사용할 수 있는 유용한 정보 (mac의 '콘솔'앱에는 찍히지 않고 xcode console에만 표출)
- Info(Info): 문제 해결 (트러블슈팅) 시 활용할 수 있는, 도움이 되지만 필수적이지 않은 정보
- warning(Notice) : 문제 해결에 필수적인 정보. Failure를 초래할 수 있는 정보 (기본값)
- Error(Error): 코드 실행 중 나타난 에러. 활동 객체가 존재하는 경우 관련 프로세스 체인에 대한 정보 캡처
- Fatal(Fault): 코드 속의 폴트 (Faults) 및 버그 관련 정보. 활동 객체가 존재하는 경우 관련 프로세스 체인에 대한 정보 캡처

```swift
    var osLogType: OSLogType {
        switch self {
        case .debug:
            return .debug
        case .info:
            return .info
        case .warning:
            return .default
        case .error:
            return .error
        case .fatal:
            return .fault
        }
    }
```

🤔
**요구사항에 있는대로 이름 바꿔서 매핑 했는데 이름이 달라서 헷갈릴까요?
Mac의 `콘솔`앱에는 시스템에 정의된 레벨이 뜨고 Xcode의 콘솔에는 직접 정의한 레벨이 뜹니다.
의견 있다면 말해주세요!**

<br>

### 로그 태그

```swift
public enum LogTagType: String {
    case view
    case viewModel
    case repository
    case network
    case auth
    case user
}
```

🤔
**OSLog를 사용하면 Mac의 `콘솔`앱에서 로그를 볼 수 있는데 태그로 필터링을 할 수 있습니다.   
일단은 이렇게 했는데 수정 할 것 있으면 알려주세요!**

<br>

### 로그 포맷
`로그레벨` `파일명` `라인넘버` `함수명` ▶︎ `메시지`

![image](https://github.com/user-attachments/assets/46cb9330-2244-4ef5-83e2-9914656456c2)

hoBahk committed:
[#15] Feat: Logger 모듈 구현

namsoo5 commented on Projects/Core/HGLogger/Sources/LogLevelType.swift:
```diff
@@ -0,0 +1,68 @@
+//
+
+import Foundation
+import os.log
```
import Logger 도 가능해서 iOS14+ 최신 로거타입이라서 참고! 
기능은똑같아서 그냥써도될듯??
https://developer.apple.com/documentation/os/logger

hoBahk replied:
아하! 그럼 함 써보겠습니다.

iosdevSW reviewed (dismissed):
굿굿 박부장님

로그 태그는 필요시 추가/제거/수정하면 되니 괜찮을 것 같슴당!

고생하셨습니다 👍

namsoo5 commented on Projects/Core/HGLogger/Sources/Loggable.swift:
```diff
@@ -0,0 +1,39 @@
+     )
+}
+
+public extension Loggable {
+    func log(
+        _ message: String,
+        level: LogLevelType,
+        tag: LogTagType? = nil,
+        file: String = #file,
+        line: Int = #line,
+        function: String = #function
+    ) {
+        log(
+            message,
+            level: level,
+            tag: tag,
+            file: file,
+            line: line,
+            function: function
+        )
+    }
+}
```
불필요해 보이는 코드인거같은데
삭제해줄수있나용?!

hoBahk replied:
프로토콜 정의 부분에는 기본값을 설정할 수 없어서 외부 모듈에서 프로토콜을 통해 함수를 사용할 때 file, line, function을 사용해야 하더라구요 그래서 넣었습니다!

namsoo5 replied:
아 맞네 그부분을 놓친거같네여 😱

namsoo5 commented on Projects/Core/HGLogger/Sources/Logger.swift:
```diff
@@ -6,3 +6,41 @@
 import Foundation
+import os.log
+
+public final class Logger: Loggable {
```
1. 시스템타입인 Logger타입이랑 겹쳐서 
네이밍 `HGLogger` 제안해봅니다

2. class이여야 하는 이유가있을까요?!
시스템 기본타입 Logger를 예시로 
struct로 인스턴스를 생성에 비용을 줄이고 필요한곳에서 생성해서 바로 로그를 찍는형식이 있을수 있어요(로그는 간편하게찍는게 최고일듯하여..)
``` swift
// 예시
let logger = Logger()
logger.log("로그입니다")
```

또는 static을이용해서 네임스페이스만 빌려와서 로그를찍는 방식도 있을거같아요
``` swift
Logger.log("로그입니다)
```

namsoo5 replied:
3. class를 제거한다면 struct에서 protocol을 채택하게되면 any를 사용할거같은데
또다른 Loggable을 준수할 타입이 생기지않을거같긴해서
값타입의 성능적 이점을 포기하고 프로토콜을 준수해야할까 라는 생각이 들어서 의견남겨봐요
-> 프로토콜 제거하는게 깔끔하려나? 에대한 생각

hoBahk replied:
1. 수정하겠습니다~

2. struct가 좋은 것 같습니다~ 수정할게요!

3. 그럼 프로토콜을 제거하고 Logger를 직접 쓰자는 뜻이지요??

namsoo5 replied:
생각해보고 의도한데로 가면될듯!!
내생각을 말한거라!!

namsoo5 commented on Projects/Core/HGLogger/Sources/Logger.swift:
```diff
@@ -6,3 +6,41 @@
+
+public final class Logger: Loggable {
+    private let subsystem: String
+    private let minLevel: LogLevelType
```
안쓰이는거같아서 제거해도좋을거같아요

hoBahk replied:
minLevel을 정해서 Logger.log() 함수 시작 할 때 minLevel 보다 우선 순위가 낮으면 로그를 찍지 않도록 했습니다!

```swift
    public func log(
        _ message: String,
        level: LogLevelType,
        tag: LogTagType? = nil,
        file: String = #file,
        line: Int = #line,
        function: String = #function
    ) {
        #if DEBUG
        guard level >= minLevel else {    <------ 사용중
            return
        }

        let filename = (file as NSString).lastPathComponent
        let category = (tag?.rawValue.capitalized) ?? "Default"
        let logger = os.Logger(subsystem: subsystem, category: category)
        
        logger.log(
            level: level.osLogType,
            "\(level.prefix) \(filename):\(line) \(function) ▶︎ \(message, privacy: .public)"
        )
        #endif
    }
```

namsoo5 replied:
헉 이부분도 놓쳣네 왜못봣지 죄송함다 😱

근데 이거 인스턴스 생성할때마다 MIN정의하고 로그를 안찍어야하는 이유가있는건가?? (궁금쓰)
LEVEL있어가지고 필터링걸고 필터될거같아보여서!!

hoBahk replied:
<img width="592" alt="image" src="https://github.com/user-attachments/assets/3d196b66-979b-4163-a3e3-9bee7d1175cf" />

요구사항에 있어서 이렇게 하긴했는데 이 요구사항은 OSLog 안썼을 때 요구사항인거 같긴해 
일단 뭐 디폴트 넣어놨으니까 냅둘게영!

namsoo5 commented on Projects/Core/HGLogger/Sources/Logger.swift:
```diff
@@ -6,3 +6,41 @@
+
+    public func log(
+        _ message: String,
+        level: LogLevelType,
```
```suggestion
        level: LogLevelType = .debug
```
같이 디폴트하나주면 좀더 편하게 쓸수잇을거같아용

hoBahk replied:
넵 반영하겠습니당

namsoo5 commented on Projects/Core/HGLogger/Sources/LogLevelType.swift:
```diff
@@ -0,0 +1,68 @@
+        }
+    }
+
+    public static func < (lhs: LogLevelType, rhs: LogLevelType) -> Bool {
```
이거 궁금한데
Comparable가 어디에 쓰이나용??

hoBahk replied:
LogLevelType이 Comparable을 준수하게 해서 minLevel이랑 비교해 우선순위가 낮으면 로그를 찍지 않도록 하는데 쓰입니다!

namsoo5 commented on Projects/Core/HGLogger/Sources/Logger.swift:
```diff
@@ -6,3 +6,41 @@
+    }
+
+    public func log(
+        _ message: String,
```
현재 로그가 String타입만 받고있긴한데
```suggestion
    _ message: @autoclosure () -> T
    
    let message = String(reflecting: message())
```
방식으로 오브젝트자체를 던질수도있긴해!! 
근데 `"\(object)"를 넣어도 똑같긴해서 
편한대로 고민해보면 좋을듯해서 의견남겨용

namsoo5 replied:
> 인스턴스 

HGLogger().log("\(object)", level: .warning) // 수정전
HGLogger().log(object, level: .warning) // 수정후 

> 문자열

HGLogger().log("문자열", level: .warning) // 수정전
HGLogger().log("문자열", level: .warning) // 수정후

쓰는차이는 이렇게다를듯

hoBahk replied:
음 고민해보겠습네다~

namsoo5 reviewed (changes requested):
굳굳 고생해쓰~~!

hoBahk committed:
[#15] feat: 피드백 반영

- Loggable 삭제
- class HGLogger -> struct HGLogger
- 오브젝트 받는 log 함수 추가
- import os.log -> import OSLog

hoBahk commented:
안녕하세요
피드백 반영했습니다~

1. class HGLogger -> struct HGLogger
    - class일 이유가 없을 것 같아서 struct로 변경했습니다.

2. Loggable 제거
    - 값타입의 성능적 이점을 가져가는게 더 좋다고 생각했습니다!

3. 오브젝트로도 받고, 문자열로도 받으면 좋을 것 같아 오브젝트 받는 함수 추가했습니다.

4. import os.log -> import OSLog
    - import Logger 찾아 봤는데 없는거 같아서 OSLog로 변경했습니다!

namsoo5 commented on Projects/Core/HGLogger/Sources/HGLogger.swift:
```diff
@@ -6,23 +6,43 @@
+    ) {
+        let message = String(reflecting: object())
+        
+        log(
```
요함수 내부적으로만 쓸거면
private처리만 해주시면 감사할게용 ㅎㅎ

hoBahk replied:
String도 받을 수 있게 두 개 모두 public으로 한건데 오브젝트만 받아도 될까요?

namsoo5 reviewed (dismissed):
고생해쑤~~

iosdevSW reviewed (dismissed):
수고 많으셨습니다 행넴!! 🙇🏻

GREENOVER commented on Projects/Core/HGLogger/Sources/HGLogger.swift:
```diff
@@ -0,0 +1,66 @@
+        
+        let filename = (file as NSString).lastPathComponent
+        let category = (tag?.rawValue.capitalized) ?? "Default"
+        let logger = Logger(subsystem: subsystem, category: category)
```
p4) 내부적으로 시스템 리소스를 사용하는 객체인데, category마다 인스턴스 새로 생성하는 구조로 보이는데 그럼 매번 로그 호출 시 Logger를 새로 만드는것에 이점이 있나요? 🤔
동일 tag로 로깅 시 해당 인스턴스를 재사용하도록 구현되어도 좋지 않을까 싶어 남겨봅니다. 
캐싱을 할 수도 있고 기타 등등일 수 있지 않을까해요.
사소한 의견입니다~

namsoo5 replied:
![image](https://github.com/user-attachments/assets/338dc496-3702-480d-9240-dc2fb71220f5)

GREENOVER replied:
그냥 하나만 더 의견드려보자면, 디버깅을 위한 로그 심는것 너무 좋은데 실 사용 환경에서도 필요한 로그 트래킹 시스템도 구축하면 좋을것 같아요!
그걸 파일로도 저장해서 떨궈주고 뭐 그런..?
지금은 사실상 개발자를 위한거니까요.
(내가 하는 일 아니라고 걍 던지기)

iosdevSW replied:
귀하신분이 오셨다 🫡🫡  안녕하세요!

namsoo5 replied:
병호가 다해줄거야 기대해

hoBahk replied:
좋은 의견 너무 감사해요~!! 👍👍
고민고민 해보겠습니다~ ㅎㅎㅎ

GREENOVER reviewed (dismissed):
안녕하세요~
지나가다가 들려봅니다.

hoBahk committed:
[#15] Feat: LoggerCache, LoggerUtil 구현

iosdevSW reviewed (approved):
나이수 나이수 😎
